Add indent/dedent patterns

piechologist opened the pull request:
Carrying over the discussion from [Zed Extensions #275](https://github.com/zed-industries/extensions/issues/275).

Indentation works well when typing the following code blocks:

```fish
function testing -d Blah
    begin
        begin
            ls
        end
    end
    
    if true
        echo
    else if true
        echo
    else
        ls
    end
    
    switch $argv[1]
    case a
        ls
    case b
        ls
    end
    
    for i in (seq 3)
        echo $i
    end
    
    while x
        ls
    end
end
``` 

The following examples can't be covered:

```fish
for i in (seq 3); echo $i; end
   # wrongly indents here 

echo "x x x" \
    "hi"
   # does not dedent here 
```

Generally, these patterns don't work when moving lines up and down with Ctrl-Cmd-Up/Down.

I think the patterns work well enough in the majority of cases. Good you put them in in the first place -- I hadn't seen them before.

Alternatively, tree-sitter has its own indentation system but I won't dive into that 😱.

piechologist committed:
Add indent/dedent patterns

piechologist commented:
BTW: I've got most things working in `highlights.scm`. Will test that in the next few days and make another PR.

piechologist commented:
Okay, I accidentally stumbled over the native tree-sitter indentation. Guess that is the proper way to implement this. Will make a new PR later today.

hasit commented:
@piechologist 

Check [nvim-treesitter's fish queries](https://github.com/nvim-treesitter/nvim-treesitter/tree/master/queries/fish) while you are at it as well. I was planning to get those in but I will hold off since you already have it working.

piechologist commented:
Yes, that repo seems to be the main target of tree-sitter-fish. 

Zed's tree view looks a bit different from what I'd expect from the grammar. 

I'm waiting for the new preview/stable versions today, will test again and make the PR with indentation and some other fixes in highlights.scm.

hasit commented:
@piechologist 

That sounds great! Thanks for taking it on.

piechologist commented:
That was pretty disappointing. 

[Nvim's indents.scm](https://github.com/nvim-treesitter/nvim-treesitter/blob/master/queries/fish/indents.scm) uses captures like `@indent.begin` and `@indent.branch` but Zed understands `@indent` and `@outdent` only. Branches are required for `else` clauses because we must outdent the line with `else` and indent the following line. Of course, putting `else` in both `@indent` and `@outdent` won't get us anywhere. 

If you want to play with it yourself, just copy nvim's indents.scm into `~/Library/Application Support/Zed/extensions/installed/fish/languages/fish/` and restart Zed. No other files need to be altered. Unfortunately, if you edit indents.scm, Zed seems to keep some old content even after reloading the extensions. So, you need to restart Zed after each edit. I'm glad it restarts so quickly!

So, I'm reopening this PR. The regex patterns may be less elegant but at least do work pretty well. 

I don't know how familiar you are with tree-sitter, I just started reading into this a few days ago. Looking forward to your comments. Then, I would prepare another PR for `highlights.scm`.

piechologist committed:
Move "and", "or", "not", and "!" from @keyword to @operator because fish
marks all these as operators too.

Also, "&&" and "||" are already in @operator and we should be consistent.

piechologist committed:
Add (list_element_access) to @constant.

It contains "..", "[", and "]" and highlights a variable expansion
like $argv[3..-4] the way fish does.

Then, ".." can be removed from @operator and the brackets can be
removed from @punctuation.bracket.

piechologist committed:
@punctuation.bracket should only highlight the parentheses and the "$"
in command substitution as fish itself does not use them for anything
else. Example: `test "$(pwd)" = "$HOME"`.

Brackets are only used for indexing in variable expansion (see the
previous commit).

Braces and commas are only used in brace expansion and fish highlights
these as operators. Tilde expansion and globbing are operators too.

piechologist committed:
Most keywords should use the predefined statements of the grammar.
This is how nvim handles it.

Add the semicolon and the ampersand to the keywords. Fish uses
fish_color_end for highlighting but we don't have a syntax token
like that so @keyword should look good and make these stand out.

piechologist committed:
We need two workarounds. See the details in highlights.scm.

piechologist committed:
Add the ERROR node. This is pretty cool. For instance, type
`if if` in a fish file and the first `if` will be highlighted
with the "hint" color.

piechologist commented:
Oops, I thought the new commits would go into a new PR.

Anyway, the new changes are in `highlights.scm` and affect only the syntax coloring. I split them into several commits. I tried to mimic fish's own syntax coloring in the terminal as closely as possible.

There's one more thing. There are two code blocks in `highlights.scm` (lines 21..29). They are supposed to highlight the operators like `-eq` in the `test` and `]` commands but don't work because the field `argument:` is missing in the grammar. Also, fish does not highlight those. We can either keep or delete these lines.

hasit commented:
@piechologist 

Thanks for the changes once again. Merged it for now since it works. We can always fix things as we go along.

For the regex pattern, we can keep it at this for now until zed adds support for reading regex with backreferencing.

For the `highlights.scm` changes, I don't see operators `not`, `and`, etc. being highlighted (check screenshot below).

<img width="1624" alt="Screenshot 2024-03-07 at 11 52 22" src="https://github.com/hasit/zed-fish/assets/1553055/8dcbb127-4e0e-4896-8a70-2346738ed857">

hasit commented:
Scratch that, it was a problem with the theme I am using that doesn't highlight operators well.

piechologist commented:
Thanks for merging! I'll keep an eye on tree-sitter-fish so we can make adjustments in case they improve their grammar.
